### ISSUE TITLE
Preload modules modules based on canLoad guard result

### ISSUE THREAD

**Issue by aciccarello** [angular/angular#20994]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Currently, when using a preloadering strategy (like `PreloadAllModules`), the preloader does not process routes with a `canLoad` property and therefore does not preload them. There is no indication in the [`CanLoad` interface documentation](https://angular.io/api/router/CanLoad) that this will happen.

https://github.com/angular/angular/blob/0f5c70d563b6943623a5940036a52fe077ad3fac/packages/router/src/router_preloader.ts#L112-L114

The [router guide](https://angular.io/guide/router#canload-blocks-preload) does note this behavior and that it is by design. However, it does not give any indication as to why.

> The PreloadAllModules strategy does not load feature areas protected by a CanLoad guard. This is by design.

The language in the guide also isn't very clear. When I originally read that part of the guide, I misunderstood "areas protected by a CanLoadGuard" as only referring to areas where the `canLoad` guard check returned false.

## Expected behavior

Preloaders should decide whether to preload based on the boolean result of the `canLoad` guard (or the boolean result of the observable/promise).

If this is not possible, it should be more clear that `canLoad` negates any preloading strategy for that module. This could be through console warnings and/or better documentation.

## What is the motivation / use case for changing the behavior?

- The `canLoad` guard allows preventing navigation before a lazy loaded module is downloaded/bootstraped
- Preloading requires "loading" a module, therefore a `canLoad` option still makes logical sense
- The suggested use of `canActivate` over `canLoad` when preloading wastes network/memory/cpu resources
- The disabling of preloading by the `canLoad` option happens silently
- `canLoad` returns an observable or promise so it can delay loading for an async task (i.e. user login)
- custom preloading strategies are slightly harder to implement than a `canLoad` hook
- custom preloading can only be configured on a root or lazy loaded module level
- `PreloadAllModules` could be used in concert with `canLoad` for a simple selective loading strategy

## Environment

<pre><code>
Angular version: 5.0.3

Browser:
- [x] Chrome (desktop) version 63.0.3239.84 (Official Build) (64-bit)
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [x] IE version 11
- [ ] Edge version XX
 
For Tooling issues:
- Node version: v8.9.1  
- Platform: Windows 7 64-bit 

Others:

- Angular CLI: 1.5.4
</code></pre>

**Comment by poddarkhushbu07** [comment]
Yes, I had the same issue. I needed to apply my authentication **CanLoad** guard and preload my module. 
This feature is currently not supported. But I applied a workaround in my application. I used _Custom Pre-loading Strategy_ which includes similar condition  available in **CanLoad** guard along.

**Comment by hodossy** [comment]
Old issue, but I have found myself in a situation where this is exactly what I need. I work a project where additional modules can be purchased and the app is deployed on-premise. The modules are implemented as lazy-loaded routes, we should only pre-load the module when it is needed, i.e. the customer purchased it (checked via an HTTP request). I still however want to pre-load the modules when needed, so the `CanActivate` trick is not an option. I managed to make this work via a custom `RouterPreloader`. This is a real pain to implement outside of Angular due to some classes are not part of the public api, therefore I think it is not really possible as a third party lib within the Angular ecosystem.

*canload_router_preloader.ts*

```typescript
import { Compiler, Injectable, Injector, NgModuleFactoryLoader, NgModuleRef, OnDestroy } from '@angular/core';
import {
  NavigationEnd,
  PreloadingStrategy,
  Route,
  RouteConfigLoadEnd,
  RouteConfigLoadStart,
  Router,
  Routes,
  UrlSegment,
} from '@angular/router';
import { prioritizedGuardValue } from '@angular/router/esm2015/src/operators/prioritized_guard_value';
import { wrapIntoObservable } from '@angular/router/esm2015/src/utils/collection';
import { isCanLoad, isFunction } from '@angular/router/esm2015/src/utils/type_guards';
import { RouterConfigLoader, LoadedRouterConfig } from '@angular/router/esm2015/src/router_config_loader';
import { EMPTY, from, Observable, of, Subscription } from 'rxjs';
import { concatMap, filter, map, mergeAll, mergeMap, switchMap } from 'rxjs/operators';

@Injectable()
export class CanLoadRouterPreloader implements OnDestroy {
  private loader: RouterConfigLoader;
  private subscription?: Subscription;

  constructor(
    private router: Router,
    moduleLoader: NgModuleFactoryLoader,
    compiler: Compiler,
    private injector: Injector,
    private preloadingStrategy: PreloadingStrategy,
  ) {
    const onStartLoad = (r: Route) => (router as any).triggerEvent(new RouteConfigLoadStart(r));
    const onEndLoad = (r: Route) => (router as any).triggerEvent(new RouteConfigLoadEnd(r));

    this.loader = new RouterConfigLoader(moduleLoader, compiler, onStartLoad, onEndLoad);
  }

  setUpPreloading(): void {
    this.subscription = this.router.events
      .pipe(
        filter((e) => e instanceof NavigationEnd),
        concatMap(() => this.preload()),
      )
      .subscribe(() => {});
  }

  preload(): Observable<any> {
    const ngModule = this.injector.get(NgModuleRef);
    return this.processRoutes(ngModule, this.router.config);
  }

  /** @nodoc */
  ngOnDestroy(): void {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  private processRoutes(ngModule: NgModuleRef<any>, routes: Routes): Observable<void> {
    const res: Observable<any>[] = [];
    for (const route of routes) {
      // we already have the config loaded, just recurse
      if (route.loadChildren && !route.canLoad && (route as any)._loadedConfig) {
        const childConfig = (route as any)._loadedConfig;
        res.push(this.processRoutes(childConfig.module, childConfig.routes));

        // no config loaded, fetch the config
      } else if (route.loadChildren) {
        if (!route.canLoad) {
          res.push(this.preloadConfig(ngModule, route));
        } else {
          res.push(
            this.runCanLoadGuards(ngModule.injector, route, []).pipe(
              switchMap((guardVal) => {
                return guardVal ? this.preloadConfig(ngModule, route) : EMPTY;
              }),
            ),
          );
        }

        // recurse into children
      } else if (route.children) {
        res.push(this.processRoutes(ngModule, route.children));
      }
    }
    return from(res).pipe(
      mergeAll(),
      map((_) => void 0),
    );
  }

  private preloadConfig(ngModule: NgModuleRef<any>, route: Route): Observable<void> {
    return this.preloadingStrategy.preload(route, () => {
      const loaded$ = (route as any)._loadedConfig
        ? of((route as any)._loadedConfig)
        : this.loader.load(ngModule.injector, route);
      return loaded$.pipe(
        mergeMap((config: LoadedRouterConfig) => {
          (route as any)._loadedConfig = config;
          return this.processRoutes(config.module, config.routes);
        }),
      );
    });
  }

  private runCanLoadGuards(moduleInjector: Injector, route: Route, segments: UrlSegment[]): Observable<boolean> {
    const canLoad = route.canLoad;
    if (!canLoad || canLoad.length === 0) return of(true);

    const canLoadObservables = canLoad.map((injectionToken: any) => {
      const guard = moduleInjector.get(injectionToken);
      let guardVal;
      if (isCanLoad(guard)) {
        guardVal = guard.canLoad(route, segments);
      } else if (isFunction(guard)) {
        guardVal = guard(route, segments);
      } else {
        throw new Error('Invalid CanLoad guard');
      }
      return wrapIntoObservable(guardVal);
    });

    return of(canLoadObservables).pipe(
      prioritizedGuardValue(),
      map((result) => result === true),
    );
  }
}
```

The idea is essentially to run the `CanLoad` guards almost exactly as found in the `apply_redirect.ts` file, except that we don't care if there is a redirect, thus the `tap` operator is deleted and `UrlTree` results are considered as `false`.

I would like to propose a new `RouterPreloader` implementation (like above, but properly implemented), for which developers would be able to opt-in by providing it:

*app.module.ts excerpt*
```typescript
  providers: [
    {
      provide: RouterPreloader,
      useClass: CanLoadRouterPreloader,
    },
  ],
```

As an alternative, I can imagine a CanPreload guard, which would have the following interface and would be respected by the default `RouterPreloader`:

*can_preload.ts*
```typescript
export interface CanPreLoad {
  canPreLoad(route: Route):
      Observable<boolean>|Promise<boolean>|boolean;
}

export type CanPreLoadFn = (route: Route) =>
    Observable<boolean>|Promise<boolean>|boolean;
```

What is your view on my proposals? Which approach is best here? I would probably go with the second option, but I needed a solution now, so that is why I implemented another `RouterPreloader`.

**Comment by atscott** [comment]
Closing since `canLoad` guards are deprecated and this behavior is not the same with `canMatch` guards in that preloading is up to the strategy rather than skipped entirely. Executing the guards, however, would still be difficult. These aren't side-effect free so I don't think we'd ever build something into the Router directly that would execute them. Different approaches could be taken in the route config itself to coordinate with a custom preloading strategy for this though.
